File: csharp/Heap.cs
/* Pairing heap */

using System;

public class Heap<T> where T : IComparable {
  Node<T> Root;

  public T Top() {
    return Root.Key;
  }

  public void Pop() {
    Root = Root.MergedSons();
  }

  public void Insert(T key) {
    Node<T> singleton = new Node<T>(key);
    Root = singleton.LinkTo(Root);
  }

  public bool IsEmpty() {
    return Root == null;
  }
}

File: java/Heap.java
/* Pairing heap */
public class Heap<T extends Comparable> {
  Node<T> root;

  public T top() {
    return root.getKey();
  }

  public void pop() {
    root = root.mergedSons();
  }

  public void insert(T key) {
    Node<T> singleton = new Node<T>(key);
    root = singleton.linkTo(root);
  }

  public boolean isEmpty() {
    return root == null;
  }
}
